code/modules/station_goals/shield.dm

```diff
@@ -0,0 +1,193 @@
+//Station Shield
+// A chain of satellites encircles the station
+// Satellites be actived to generate a shield that will block unorganic matter from passing it.
+/datum/station_goal/station_shield
+	name = "Station Shield"
+	var/coverage_goal = 500
+
+/datum/station_goal/station_shield/get_report()
+	return {"<b>Station Shield construction</b><br>
+	The station is located in a zone full of space debris. We have a prototype shielding system you will deploy to reduce collision related accidents.
+	<br><br>
+	You can order the satellites and control systems through cargo shuttle."}
+
+/datum/station_goal/station_shield/on_report()
+	//Unlock
+	var/datum/supply_packs/P = shuttle_master.supply_packs["[/datum/supply_packs/misc/shield_sat]"]
+	P.special_enabled = TRUE
+
+	P = shuttle_master.supply_packs["[/datum/supply_packs/misc/shield_sat_control]"]
+	P.special_enabled = TRUE
+
+/datum/station_goal/station_shield/check_completion()
+	if(..())
+		return TRUE
+	if(get_coverage() >= coverage_goal)
+		return TRUE
+	return FALSE
+
+/datum/station_goal/proc/get_coverage()
+	var/list/coverage = list()
+	for(var/obj/machinery/satellite/meteor_shield/A in machines)
+		if(!A.active || !is_station_level(A.z))
+			continue
+		coverage |= view(A.kill_range, A)
+	return coverage.len
+
+/obj/item/weapon/circuitboard/computer/sat_control
+	name = "Satellite Network Control (Computer Board)"
+	build_path = /obj/machinery/computer/sat_control
+	origin_tech = "engineering=3"
+
+/obj/machinery/computer/sat_control
+	name = "Satellite control"
+	desc = "Used to control the satellite network."
+	circuit = /obj/item/weapon/circuitboard/computer/sat_control
+	icon_screen = "accelerator"
+	icon_keyboard = "accelerator_key"
+	var/notice
+	
+/obj/machinery/computer/sat_control/attack_hand(mob/user)
+	if(..())
+		return 1
+	ui_interact(user)
+
+/obj/machinery/computer/sat_control/ui_interact(mob/user, ui_key = "main", var/datum/nanoui/ui = null, var/force_open = 1)
+	ui = nanomanager.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "sat_control.tmpl", name, 475, 400)
+		ui.open()
+
+/obj/machinery/computer/sat_control/Topic(href, href_list)
+	if(..())
+		return 1
+
+	if(href_list["toggle"])
+		toggle(text2num(href_list["id"]))
+		. = TRUE
+
+/obj/machinery/computer/sat_control/proc/toggle(id)
+	for(var/obj/machinery/satellite/S in machines)
+		if(S.id == id && S.z == z)
```

Comment:
Could the `S.z == z` be replaced with `atoms_share_level(src, S)`, so it's easier to extend in the future?